File: assignments/submissions/40250382/02/operacoes_matrizes.js
/**
 * Assignment 02: Matrix Operations and Property Verification
 * Tarefas T1 a T5
 */

// Se estiveres a usar Node.js, descomenta a linha abaixo:
// const math = require('mathjs');

console.log("%c--- Task 1: Operações Básicas (A+B, A-B, 3A) ---", "color: blue; font-weight: bold");
const A = math.matrix([[1, 2, 3], [4, 5, 6]]);
const B = math.matrix([[7, 8, 9], [10, 11, 12]]);

const sum = math.add(A, B);
const diff = math.subtract(A, B);
const scalar = math.multiply(3, A);

console.log("A + B:", sum.toArray());
console.log("A - B:", diff.toArray());
console.log("3A:", scalar.toArray());

// Verificação de dimensões usando math.size e assertion
const sizeA = math.size(A).toArray();
const sizeSum = math.size(sum).toArray();
console.assert(JSON.stringify(sizeA) === JSON.stringify(sizeSum), "Erro: Dimensões não coincidem!");
console.log("Verificação: Dimensões preservadas na soma.");

console.log("\n%c--- Task 2: Multiplicação AC vs CA (Análise Dimensional) ---", "color: blue; font-weight: bold");
const C = math.matrix([[1, 2], [3, 4], [5, 6]]); // 3x2

const AC = math.multiply(A, C); // (2x3) * (3x2) = 2x2
const CA = math.multiply(C, A); // (3x2) * (2x3) = 3x3

console.log("AC (2x2):", AC.toArray());
console.log("CA (3x3):", CA.toArray());
console.log("Explicação: AC é 2x2 e CA é 3x3. A multiplicação de matrizes não é comutativa e o tamanho do resultado depende da ordem (linhas de uma vs colunas de outra).");

console.log("\n%c--- Task 3: Comutador e Identidade ---", "color: blue; font-weight: bold");
const P = math.matrix([[1, 2], [3, 4]]);
const Q = math.matrix([[5, 6], [7, 8]]);
const I = math.identity(2);

const PQ = math.multiply(P, Q);
const QP = math.multiply(Q, P);
const commutator = math.subtract(PQ, QP);

console.log("PQ:", PQ.toArray());
console.log("QP:", QP.toArray());
console.log("Comutador [P, Q] = PQ - QP:", commutator.toArray());
console.log("P * I = P:", math.deepEqual(math.multiply(P, I), P));

console.log("\n%c--- Task 4: Propriedades da Transposta ---", "color: blue; font-weight: bold");
// Verificações com arredondamento para evitar erros de precisão float
const round = (m) => math.round(m, 10);

// (a) (A^T)^T = A
const propA = math.deepEqual(round(math.transpose(math.transpose(A))), round(A));
// (b) (A+B)^T = A^T + B^T
const propB = math.deepEqual(round(math.transpose(math.add(A, B))), round(math.add(math.transpose(A), math.transpose(B))));
// (c) (AC)^T = C^T * A^T
const propC = math.deepEqual(round(math.transpose(AC)), round(math.multiply(math.transpose(C), math.transpose(A))));

console.log("(A^T)^T = A:", propA);
console.log("(A+B)^T = A^T + B^T:", propB);
console.log("(AC)^T = C^T * A^T:", propC);

console.log("\n%c--- Task 5: Simetria e Combinação Linear ---", "color: blue; font-weight: bold");
// Matriz Simétrica
const M = math.random([3, 3]);
const S = math.add(M, math.transpose(M));
console.log("S = M + M^T é simétrica:", math.deepEqual(round(S), round(math.transpose(S))));

// Ax como Combinação Linear das Colunas
const SysA = math.matrix([[1, 2, 3], [0, 1, 4], [5, 6, 0]]);
const xVec = [2, -1, 3];
const Ax = math.multiply(SysA, xVec);

const a1 = math.flatten(math.column(SysA, 0));
const a2 = math.flatten(math.column(SysA, 1));
const a3 = math.flatten(math.column(SysA, 2));

const linearCombination = math.add(
    math.multiply(xVec[0], a1),
    math.add(math.multiply(xVec[1], a2), math.multiply(xVec[2], a3))
);

console.log("Resultado Ax:", Ax.toArray());
console.log("Resultado via Combinação Linear:", linearCombination.toArray());
console.log("Os resultados são iguais:", math.deepEqual(round(Ax), round(linearCombination)));

File: assignments/submissions/40230223/02/python/operacoes_matrizes.py
import numpy as np


def print_section(title):
    print(f"\n----------- {title} -----------\n")


def show_matrix(name, matrix):
    print(f"{name}:")
    print(matrix)
    print("\n")


# ------------------ T1 ------------------
print_section("T1")

# definição das matrizes
A = np.array([[1, 2, 3], [4, 5, 6]])
B = np.array([[6, 5, 4], [3, 2, 1]])

show_matrix("Matrix A", A)
show_matrix("Matrix B", B)

# operações básicas
sum_ab = A + B
diff_ab = A - B
scalar_a = 3 * A

show_matrix("A + B", sum_ab)
show_matrix("A - B", diff_ab)
show_matrix("3 * A", scalar_a)

assert sum_ab.shape == A.shape == B.shape
print("Depois da soma, a dimensão mantém-se igual.\n")


# ------------------ T2 ------------------
print_section("T2")

C = np.array([[1, 2], [3, 4], [5, 6]])
show_matrix("Matrix C", C)

# multiplicação
AC = A @ C
show_matrix("A @ C", AC)
print(f"Dimensão de AC: {AC.shape}\n")

CA = C @ A
show_matrix("C @ A", CA)
print(f"Dimensão de CA: {CA.shape}\n")

print("multiplicação de matrizes não é comutativa.\n")


# ------------------ T3 ------------------
print_section("T3")

P = np.array([[1, 2], [3, 4]])
Q = np.array([[2, 0], [1, 2]])

PQ = P @ Q
QP = Q @ P

show_matrix("P @ Q", PQ)
show_matrix("Q @ P", QP)

# commutator
comm = PQ - QP
show_matrix("[P, Q] = PQ - QP", comm)

print("PQ == QP ?", np.allclose(PQ, QP), "\n")

# matriz identidade
I = np.eye(2)

print("P @ I == P ?", np.allclose(P @ I, P))
print("I @ P == P ?", np.allclose(I @ P, P), "\n")


# ------------------ T4 ------------------
print_section("T4")

prop_a = np.allclose((A.T).T, A)
prop_b = np.allclose((A + B).T, A.T + B.T)
prop_c = np.allclose((A @ C).T, C.T @ A.T)

print("(A^T)^T = A ->", prop_a)
print("(A + B)^T = A^T + B^T ->", prop_b)
print("(AC)^T = C^T A^T ->", prop_c)

# ------------------ T5 ------------------
print_section("T5")

M = np.random.rand(3, 3)
S = M + M.T

show_matrix("M", M)
show_matrix("S = M + M^T", S)

print("S simétrico - ", np.allclose(S, S.T), "\n")

# sistema 3x3
A_3 = np.array([[1, 4, 7], [2, 5, 8], [3, 6, 9]])
x = np.array([2, -1, 3])

Ax = A_3 @ x

# combinação linear das colunas
a1 = A_3[:, 0]
a2 = A_3[:, 1]
a3 = A_3[:, 2]

lin_comb = x[0] * a1 + x[1] * a2 + x[2] * a3

show_matrix("A * x", Ax)
show_matrix("x1*a1 + x2*a2 + x3*a3", lin_comb)

print("Ax == combinação linear ?", np.allclose(Ax, lin_comb))
